tests/queries/0_stateless/02458_datediff_date32.reference

```diff
@@ -0,0 +1,169 @@
+-- { echo }
+
+-- Date32 vs Date32
+SELECT dateDiff('second', toDate32('1900-01-01'), toDate32('1900-01-02'));
+86400
+SELECT dateDiff('minute', toDate32('1900-01-01'), toDate32('1900-01-02'));
+1440
+SELECT dateDiff('hour', toDate32('1900-01-01'), toDate32('1900-01-02'));
+24
+SELECT dateDiff('day', toDate32('1900-01-01'), toDate32('1900-01-02'));
+1
+SELECT dateDiff('week', toDate32('1900-01-01'), toDate32('1900-01-08'));
+1
+SELECT dateDiff('month', toDate32('1900-01-01'), toDate32('1900-02-01'));
+1
+SELECT dateDiff('quarter', toDate32('1900-01-01'), toDate32('1900-04-01'));
+1
+SELECT dateDiff('year', toDate32('1900-01-01'), toDate32('1901-01-01'));
+1
+-- With DateTime64
+-- Date32 vs DateTime64
+SELECT dateDiff('second', toDate32('1900-01-01'), toDateTime64('1900-01-02 00:00:00', 3));
+86400
+SELECT dateDiff('minute', toDate32('1900-01-01'), toDateTime64('1900-01-02 00:00:00', 3));
+1440
+SELECT dateDiff('hour', toDate32('1900-01-01'), toDateTime64('1900-01-02 00:00:00', 3));
+24
+SELECT dateDiff('day', toDate32('1900-01-01'), toDateTime64('1900-01-02 00:00:00', 3));
+1
+SELECT dateDiff('week', toDate32('1900-01-01'), toDateTime64('1900-01-08 00:00:00', 3));
+1
+SELECT dateDiff('month', toDate32('1900-01-01'), toDateTime64('1900-02-01 00:00:00', 3));
+1
+SELECT dateDiff('quarter', toDate32('1900-01-01'), toDateTime64('1900-04-01 00:00:00', 3));
+1
+SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));
+1
```

Comment:
I'm not sure is this change breaks it or not, likely it is introduced in #39425, but I will write here since anyway you are the author of both and it seems that you really use this extended support.

This line is flaky - https://play.clickhouse.com/play?user=play#c2VsZWN0IAogIHJlcG9ydF91cmwKZnJvbSBjaGVja3MKd2hlcmUKICBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gaW50ZXJ2YWwgNyBkYXkgYW5kCiAgdGVzdF9uYW1lIGxpa2UgJyUwMjQ1OF9kYXRlZGlmZl9kYXRlMzIlJyBhbmQKICB0ZXN0X3N0YXR1cyBpbiAoJ0ZBSUwnLCAnRkxBS1knLCAnU0tJUFBFRCcp

For `Pacific/Guam` timezone it has different result:

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'UTC') AS utc,
    toDateTime64('1901-01-01 00:00:00', 3, 'Pacific/Guam') AS fail

Row 1:
──────
utc:  1901-01-01 00:00:00.000
fail: 1900-12-31 23:45:00.000
```

And this is for 22.7 (without referenced PR):

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'UTC') AS utc,
    toDateTime64('1901-01-01 00:00:00', 3, 'Pacific/Guam') AS fail

Row 1:
──────
utc:  1925-01-01 00:00:00.000
fail: 1925-01-01 00:00:00.000
```

Also I don't understand how DateTime64 equality works, is this a bug?

```sql
SELECT toDateTime('1901-01-01 00:00:00', 'UTC') = toDateTime('1901-01-01 00:00:00', 'Europe/Moscow') AS eq

┌─eq─┐
│  1 │
└────┘

SELECT toDateTime64('1901-01-01 00:00:00', 3, 'UTC') = toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow') AS eq

┌─eq─┐
│  0 │
└────┘
```

And one more thing, why `toDateTime64` differs from `toDateTime`:

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow'),
    toDateTime('1901-01-01 00:00:00', 'Europe/Moscow')

Row 1:
──────
toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow'): 1925-01-01 00:00:00.000
toDateTime('1901-01-01 00:00:00', 'Europe/Moscow'):      1970-01-01 03:00:00
``

Reply:
So the fix would be to specify the time zone (e.g. `UTC`) in `toDateTime64()`. I have seen the same issue leading to flakiness in other tests too.

> Also I don't understand how DateTime64 equality works, is this a bug?

About
```sql
SELECT toDateTime('1901-01-01 00:00:00', 'UTC') = toDateTime('1901-01-01 00:00:00', 'Europe/Moscow') AS eq
```
Date type `DateTime` before 1970 is undefined, hence the comparison is meaningless.

`SELECT toDateTime64('1901-01-01 00:00:00', 3, 'UTC') = toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow') AS eq` gives the expected result (`0`).

> And one more thing, why toDateTime64 differs from toDateTime:

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow')
    toDateTime('1901-01-01 00:00:00', 'Europe/Moscow')
```

This gives on my system

```
┌─toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow')─┬─toDateTime('1901-01-01 00:00:00', 'Europe/Moscow')─┐
│                                 1901-01-01 00:00:00.000 │                                1970-01-01 03:00:00 │
└─────────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘
```

Reply:
@azat I made a series of PRs related to adding extended range support for Date/Time functions. But I did no modify toDateTime64 and toDateTime functions in my PRs.
I will take a look at it today.

As for toDateTime function - yes, as @rschu1ze said: the value is undefined before 1970-01-01 so any arguments for toDate or toDateTime function before 1970 like 1900 or 1901 will be "narrowed" to 1970-01-01.